File: src/p001/p001.hs
module P001 where

listCompSolution :: Integer
listCompSolution = sum $ filter isDivisibleBy3or5 [1 .. 999]
  where
    isDivisibleBy3or5 x = mod x 3 == 0 || mod x 5 == 0

sumMultiplesOf :: Integer -> Integer
sumMultiplesOf num = numberOfMultiples * (numberOfMultiples + 1) * num `div` 2
  where
    numberOfMultiples = 999 `div` num

seriesSolution :: Integer
seriesSolution = sumMultiplesOf 3 + sumMultiplesOf 5 - sumMultiplesOf 15

main :: IO ()
main = do
  print seriesSolution
  print listCompSolution

File: src/p001/p001.scala
def listFilterSolution(): Int =
  (1 until 1000).filter { x => x % 3 == 0 || x % 5 == 0 }.sum

def sumMultiplesOf(n: Int): Int =
  val numberOfMultiples = 999 / n
  numberOfMultiples * (numberOfMultiples + 1) * n / 2

def seriesSolution(): Int =
  sumMultiplesOf(3) + sumMultiplesOf(5) - sumMultiplesOf(15)

@main def main(): Unit =
  println(listFilterSolution())
  println(seriesSolution())
